org.eclipse.ice.commands/src/main/java/org/eclipse/ice/commands/RemoteTransferExecution.java

```diff
@@ -105,16 +112,68 @@ private void transferRemoteToRemote(Connection connection, String source, String
 		// Build the command to execute
 		String command = moveType + source + " " + destination;
 		// Set the command for the JSch connection
-		execChannel.setCommand(command);
-		// If the channel isn't connected, connect and run the command
+		connection.getSession().executeRemoteCommand(command);
+	}
+	
+	/**
+	 * Transfers a local file to the remote host. If the remote destination is a directory, a file with the
+	 * same name as the local file will be created in the directory.
+	 * 
+	 * @param connection
+	 * @param src
+	 * @param dest
+	 * @throws IOException
+	 */
+	private void transferLocalToRemote(Connection connection, String src, String dest) throws IOException {
+		SftpClient client = connection.getSftpChannel();
 		try {
-			execChannel.connect();
-		} catch (JSchException e) {
-			logger.error("Channel isn't connected and can't copy remote to remote...", e);
-			throw e;
+			if (client.stat(dest).isDirectory()) {
+				Path path = FileSystems.getDefault().getPath(src);
+				String sep = "";
+				if (!dest.endsWith("/")) {
+					sep = "/";
+				}
+				dest += sep + path.getFileName();
+			}	
+		} catch (SftpException e) {
+            if (!(e.getStatus() == SftpConstants.SSH_FX_NO_SUCH_FILE)) {
+                throw e;
+            }
+        }
+		try (OutputStream dstStream = client.write(dest, OpenMode.Create, OpenMode.Write, OpenMode.Truncate)) {
+			try (InputStream srcStream = new FileInputStream(src)) {
+				byte[] buf = new byte[32*1024];
+				while (srcStream.read(buf) > 0) {
+					dstStream.write(buf);
+				}
+			}
```

Comment:
Just to check my own understanding, is this transfer code not actually copying/moving a file but rather reading it and writing it over the ssh connection through an input and output stream?

Reply:
The reason I ask is that the output files which result from executing the tests where a python script is run have a bunch of weird symbols at the end - maybe carriage returns or something like that? Nominally the tests just delete these files so as to not clutter up one's working directory, but you can see this for yourself by just commenting out the `@AfterClass` decorators in the test classes that delete files (for example, in `CommandFactoryTest.java`)

Reply:
The only way to transfer from local to remote (or remote to local) is via streams. If you look at the implementation of Jsch put() and get(), that's what it does under the covers. It's just that Mina doesn't provide a put() and get() method so you have to do it yourself. It sounds more like the issue is to do with character encoding, which I admit I didn't pay a lot of attention to, but I'll take a look at the test to see what is going on.

Reply:
Well, the thing I pointed out would have been hard to catch anyway since the tests delete the output files automatically. It only was the case for the python script output files as far as I could tell.

Reply:
Do you know if it's the same for the Jsch version?

Reply:
Sorry - thought I responded to this but apparently did not. The same thing does not happen for the JSch version, so it must be something to do with the way Mina is writing to the output files.